test/torch_optimizer_test.py

```diff
@@ -37,10 +37,49 @@ class LinearNet(nn.Module):
     def __init__(self, input_dim, output_dim):
         super(LinearNet, self).__init__()
         self.fc = nn.Linear(input_dim, output_dim)
+        self._num_parameters = input_dim*output_dim
 
     def forward(self, x):
         return self.fc(x)
 
+    @property
+    def num_parameters(self):
+        return self._num_parameters
+
+# A deep linear model for testing
+class DuplicatedLinearNet(nn.Module):
+    def __init__(self, input_dim, output_dim):
+        super(DuplicatedLinearNet, self).__init__()
+        self.fc1 = nn.Linear(input_dim, output_dim)
+        self.fc2 = nn.Linear(output_dim, output_dim)
+        self._num_parameters = input_dim*output_dim+output_dim**2
+
+    def forward(self, x):
+        x = self.fc1(x)
+        x = self.fc2(x)
+        x = self.fc2(x)
+        return x
+    
+    @property
+    def num_parameters(self):
+        return self._num_parameters
+
+class HierarchicalLinearNet(nn.Module):
+    def __init__(self, input_dim, output_dim):
+        super(HierarchicalLinearNet, self).__init__()
+        self.fc1 = LinearNet(input_dim, output_dim)
+        self.fc2 = LinearNet(output_dim, output_dim)
```

Comment:
Maybe it is better with one LinearNet and another DuplicatedLinearNet? Because LinearNet is almost same as nn.Linear.

Reply:
Discussed offline. Ignore this for now.